core/src/main/java/org/infinispan/interceptors/impl/TxInterceptor.java

```diff
@@ -301,19 +332,53 @@ public int size() {
                   return (int) size;
                }
             };
-         }
-         return rv;
-      });
+         });
+      }
+      return super.visitKeySetCommand(ctx, command);
    }
 
    @Override
    public Object visitEntrySetCommand(InvocationContext ctx, EntrySetCommand command) throws Throwable {
       enlistIfNeeded(ctx);
-      return invokeNextThenApply(ctx, command, (rCtx, rCommand, rv) -> {
-         if (rCtx.isInTxScope()) {
+      if (ctx.isInTxScope()) {
+         // We just set it, unfortunately we always wrap even if this flag was set
```

Comment:
I don't expect anyone to benchmark calls to `cache.entrySet()` without actually iterating through the set, so I don't think an extra wrapper instance matters.

Reply:
It does actually. And actually looking at this closer, this is highlighted because there is a bug in `TxInterceptor` that it isn't properly wrapping the iterator to support `remove` operation. Normally this would cause an extra wrapper per iterator instance.

Reply:
I have another test to add it looks like.

Reply:
I just meant there's no need for "unfortunately" in the comment, we need to wrap it for correctness so we do it.